File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    boolean res=false;                          // Shuru mei Global Variable ko false man k chlo
    public void fun(TreeNode root,int sum,int targetSum){
        // Agar Root hi null hai; yani ki Tree hi nhi hai
        if(root==null){
            return;
        }

        sum+=root.val;                         // Shuru mei by default sum 0 hi hai; Path k along sum kardo

        // Agar Leaf hai
        if(root.left==null && root.right==null){
            if(sum==targetSum){
                // Leaf node tak pahunch k Path Sum target k equal mil gya
                res=true;
            }
            return;
        }
        // Agar Non Leaf node hua; toh aagey k nodes Traverse karna hai
        fun(root.left,sum,targetSum);          // Left wley Child ko Traverse krtey rho
        fun(root.right,sum,targetSum);         // Right wley Child ko Traverse kro
    }

    public boolean hasPathSum(TreeNode root,int targetSum) {
        // Recursive Function Call kardo
        fun(root,0,targetSum);

        return res;                            // Final Answer return kardo
    }
}

File: 0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public: 
bool res=false;                                    // shuru mei global var ko false man k chl lo 
void fun(TreeNode*root, int sum, int targetSum){
    // Agar root hi null hai; yani ki tree hi nhi hai 
    if(root==nullptr){
        return;
    }
    sum+=root->val;                            // shuru mei by default sum 0 hi hai sum kardo path k along 

    // Agar leaf hai 
    if(root->left==nullptr && root->right==nullptr){
        if(sum==targetSum){
            res=true;                // agr leaf node tak pahunch k path sum target k equal mil gya ;ret true
        }
        return;
    }

    // agar Non leaf node hua ; toh agey k nodes traverse karna hai 
    fun(root->left,sum,targetSum);                   // left wley child ko traverse krtey rho
    fun(root->right,sum,targetSum);                  // right wley child ko trav kro 
}

    bool hasPathSum(TreeNode* root, int targetSum) {
        // Recursive Function call kardo 
        fun(root,0,targetSum);

        return res;            // agar target wla mila toh true ret hojaega 
    }
};